gdal/swig/python/osgeo/utils/auxiliary/util.py

```diff
@@ -0,0 +1,159 @@
+    if not filename:
+        return 'MEM'
+    drv_list = GetOutputDriversFor(filename, is_raster)
+    ext = get_extension(filename)
+    if not drv_list:
+        if not ext:
+            return default_raster_format if is_raster else default_vector_format
+        else:
+            raise Exception("Cannot guess driver for %s" % filename)
+    elif len(drv_list) > 1:
+        print("Several drivers matching %s extension. Using %s" % (ext if ext else '', drv_list[0]))
+    return drv_list[0]
+
+
+def open_ds(filename_or_ds: path_or_ds, *args, **kwargs):
+    ods = OpenDS(filename_or_ds, *args, **kwargs)
+    return ods.__enter__()
+
+
+def get_ovr_count(filename_or_ds: path_or_ds):
+    with OpenDS(filename_or_ds) as ds:
+        bnd = ds.GetRasterBand(1)
+        return bnd.GetOverviewCount()
+
+
+def get_ovr_idx(filename_or_ds: path_or_ds, ovr_idx: Optional[int]):
+    if ovr_idx is None:
+        ovr_idx = 0
+    if ovr_idx < 0:
+        # -1 is the last overview; -2 is the one before the last
+        overview_count = get_ovr_count(open_ds(filename_or_ds))
+        ovr_idx = max(0, overview_count + ovr_idx + 1)
+    return ovr_idx
+
+
+class OpenDS:
+    __slots__ = ['filename', 'ds', 'args', 'kwargs', 'own', 'silent_fail']
+
+    def __init__(self, filename_or_ds: path_or_ds, silent_fail=False, *args, **kwargs):
+        self.ds: Optional[gdal.Dataset] = None
+        self.filename: Optional[PathLike] = None
+        if is_path_like(filename_or_ds):
+            self.filename = str(filename_or_ds)
+        else:
+            self.ds = filename_or_ds
+        self.args = args
+        self.kwargs = kwargs
+        self.own = None
+        self.silent_fail = silent_fail
+
+    def __enter__(self) -> gdal.Dataset:
+
+        if self.ds is None:
+            self.ds = self._open_ds(self.filename, *self.args, **self.kwargs)
+            if self.ds is None and not self.silent_fail:
+                raise IOError('could not open file "{}"'.format(self.filename))
+            self.own = True
+        return self.ds
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.own:
+            self.ds = None
+
+    @staticmethod
+    def _open_ds(
+        filename: PathLike,
+        access_mode=gdal.GA_ReadOnly,
+        ovr_idx: int = None,
+        open_options: dict = None,
+        logger=None,
+    ):
+        open_options = dict(open_options or dict())
+        ovr_idx = get_ovr_idx(filename, ovr_idx)
+        if ovr_idx > 0:
+            open_options["OVERVIEW_LEVEL"] = ovr_idx - 1  # gdal overview 0 is the first overview (after the base layer)
+        if logger is not None:
+            s = 'openning file: "{}"'.format(filename)
+            if open_options:
+                s = s + " with options: {}".format(str(open_options))
+            logger.debug(s)
+        if open_options:
+            open_options = ["{}={}".format(k, v) for k, v in open_options.items()]
+
+        if open_options:
```

Comment:
How? is GDALAccess eAccess == unsigned int nOpenFlags ?

'''
def Open(*args):
    """Open(char const * utf8_path, GDALAccess eAccess) -> Dataset"""
    return _gdal.Open(*args)

def OpenEx(*args, **kwargs):
    """OpenEx(char const * utf8_path, unsigned int nOpenFlags=0, char ** allowed_drivers=None, char ** open_options=None, char ** sibling_files=None) -> Dataset"""
    return _gdal.OpenEx(*args, **kwargs)
'''

Reply:
> is GDALAccess eAccess == unsigned int nOpenFlags ?

Yeah, it happens that values were selected such that GA_ReadOnly == gdal.OF_READONLY in open flags and GA_Update == gdal.OF_UPDATE  but it would be cleaner to use gdal.OF_READONLY / OF_UPDATE though

Reply:
OK, thanks.